### PR TITLE
Upgrade `doc/make.jl` to introspect branch/tag names from git info

### PR DESCRIPTION
**Pull request by staticfloat**
Note that this requires that the julia you're using to deploy the docs matches the current julia checkout.

### CODE REVIEW

**Comment by mortenpi** [comment]
Also needs `TRAVIS_PULL_REQUEST = "false"` when building on branches. It should be fine to just set it iff on CI -- the tag and branch checks should make sure that we only deploy on non-PR builds.

**Comment by staticfloat** [comment]
What happens if `TRAVIS_PULL_REQUEST` is not set?  I was planning on just only running `make -C doc deploy` when on `master` or `release-*`.

**Comment by mortenpi** [comment]
It won't deploy. If we already check for PRs, then `TRAVIS_PULL_REQUEST` could just always be set to `false`.

The only minor this is that `deploydocs` will also try to deploy when running `make deploy` locally then (because all the env.vars are now set; well, it will fail due to a missing `DOCUMENTER_KEY` most likely). But that's probably OK.. the average user is not expected to run `make -C doc deploy` anyway I think.

File: doc/make.jl [hunk 1]
```diff
@@ -191,8 +191,19 @@ makedocs(
...
 # docs from different workers.
 if "deploy" in ARGS && Sys.ARCH === :x86_64 && Sys.KERNEL === :Linux
-# Override TRAVIS_REPO_SLUG since we deploy to a different repo
-withenv("TRAVIS_REPO_SLUG" => "JuliaLang/docs.julialang.org") do
+
+# Override a few environment variables to deploy to the appropriate repository,
+# encode things like branch, whether it's built on a tag, etc....
+env_mappings = [
+    "TRAVIS_REPO_SLUG" => "JuliaLang/docs.julialang.org",
+    "TRAVIS_BRANCH" => Base.GIT_VERSION_INFO.branch,
```
Comment (mortenpi) [inline doc/make.jl]:
```suggestion
    "TRAVIS_BRANCH" => Base.GIT_VERSION_INFO.branch,
    "TRAVIS_PULL_REQUEST" = "false",
```

**Comment by staticfloat** [comment]
> The only minor this is that deploydocs will also try to deploy when running make deploy locally then

Ah, I see. In that case, let's set `TRAVIS_PULL_REQUEST` to `false` [within the buildbot configuration](https://github.com/JuliaCI/julia-buildbot/commit/2fb4e16de00203ee7ff325e3e6786909b6e4d327).

**Comment by staticfloat** [comment]
I'm going to merge this, then we'll see if the build on `master` goes through properly.  :D

**Pull request merged**